tianshou/data/batch.py

```diff
@@ -967,3 +1135,151 @@ def split(
+        inplace: Literal[True],
+    ) -> None:
+        ...
+
+    @overload
+    def apply_values_transform(
+        self,
+        values_transform: Callable,
+        inplace: Literal[False],
+    ) -> Self:
+        ...
+
+    def apply_values_transform(
+        self,
+        values_transform: Callable,
+        inplace: bool = False,
+    ) -> None | Self:
+        """Applies a function to all non-batch-values in the batch, including
+        values in nested batches.
+
+        A batch with keys pointing to either batches or to non-batch values can
+        be thought of as a tree of Batch nodes. This function traverses the tree
+        and applies the function to all leaf nodes (i.e. values that are not
+        batches themselves).
+
+        The values are usually arrays, but can also be scalar values of an
+        arbitrary type since retrieving a single entry from a Batch a la
+        `batch[0]` will return a batch with scalar values.
+        """
+        return _apply_batch_values_func_recursively(self, values_transform, inplace=inplace)
+
+    def set_array_at_key(
+        self,
+        arr: np.ndarray,
+        key: str,
+        index: IndexType | None = None,
+        default_value: float | None = None,
+    ) -> None:
+        if index is not None:
+            if key not in self.get_keys():
+                log.info(
+                    f"Key {key} not found in batch, "
+                    f"creating a sequence of len {len(self)} with {default_value=} for it.",
+                )
+                try:
+                    self[key] = np.array([default_value] * len(self), dtype=arr.dtype)
+                except TypeError as exception:
+                    raise TypeError(
+                        f"Cannot create a sequence of dtype {arr.dtype} with default value {default_value}. "
+                        f"You can fix this either by passing an array with the correct dtype or by passing "
+                        f"a different default value that can be cast to the array's dtype (or both).",
+                    ) from exception
+            else:
+                existing_entry = self[key]
+                if isinstance(existing_entry, BatchProtocol):
+                    raise ValueError(
+                        f"Cannot set sequence at key {key} because it is a nested batch, "
+                        f"can only set a subsequence of an array.",
+                    )
+            self[key][index] = arr
+        else:
+            if len(arr) != len(self):
+                raise ValueError(
+                    f"Sequence length {len(arr)} does not match "
+                    f"batch length {len(self)}. For setting a subsequence with missing "
+                    f"entries filled up by default values, consider passing an index.",
+                )
+            self[key] = arr
+
+    def isnull(self) -> Self:
+        return self.apply_values_transform(pd.isnull, inplace=False)
```

Comment:
why do you need pandas to check isnull? is there a better way, e.g., using numpy?

Reply:
Numpy fails with certain types of NaNs (maybe None in arrays of certain types), tests failed. Pandas is more robust, that's why I used it. I'll add a block comment